src/main/java/org/hydev/mcpm/client/interaction/ProgressBar.java

```diff
@@ -18,61 +40,118 @@ public class ProgressBar implements AutoCloseable
     public ProgressBar(ProgressBarTheme theme)
     {
         this.theme = theme;
-        this.init();
-    }
+        this.out = System.out;
+        this.cu = new ConsoleUtils(this.out);
+        this.activeBars = new ArrayList<>();
+        this.cols = AnsiConsole.getTerminalWidth();
 
-    /**
-     * Initialize the progress bar (print the first line)
-     */
-    public void init()
-    {
-        // TODO: Implement this
-        throw new UnsupportedOperationException("TODO");
+        // Default to 70-char width if the width can't be detected (like in a non-tty output)
+        if (this.cols == 0) this.cols = 70;
+
+        // Last update time
+        this.lastUpdate = System.nanoTime();
+
+        // Default frame delay is 0.01666 (60 fps)
+        this.frameDelay = 1 / 60d;
+
+        // Check if output is a TTY. If not, change frame rate to 0.5 fps to avoid spamming a log.
+        this.istty = isatty(STDOUT_FILENO) == 0;
+        if (istty) this.frameDelay = 1 / 0.5;
     }
 
     /**
      * Append a progress bar at the end
      *
-     * @return Unique identifier of the progress bar
+     * @param bar Row of the progress bar
+     * @return bar for fluent access
      */
-    public String appendBar()
+    public ProgressRow appendBar(ProgressRow bar)
     {
-        // TODO: Implement this
-        throw new UnsupportedOperationException("TODO");
+        this.activeBars.add(bar);
+        bar.setPb(this);
+
+        out.println();
+        update();
+        return bar;
     }
 
-    /**
-     * Set progress for a bar
-     *
-     * @param id Unique identifier
-     * @param progress Progress as a ratio in range 0-1
-     */
-    public void setBar(String id, float progress)
+    protected void update()
+    {
+        // Check time to limit for framerate (default 60fps)
+        // Performance of the update heavily depends on the terminal's escape code handling
+        // implementation, so frequent updates will degrade performance on a bad terminal
+        var curTime = System.nanoTime();
+        if ((curTime - lastUpdate) / 1e9d < frameDelay) return;
```

Comment:
I think 1e9d is still a bit of a magic number (S_TO_PS or something). You're call if you want to factor it out.